### ISSUE TITLE
URL sharing doesn't work on iOS 18

### ISSUE THREAD

**Issue by darix795** [achorein/expo-share-intent#105]
**Describe the bug**
Yesterday I upgraded Xcode to version 16 and try to run my app on iOS 18, but seems that when I share urls to my app controller is showed but app isn't opened.
When I try on simulator with iOS 17 everything works fine. 

**To Reproduce**
Try on iOS 18 and that's it

https://github.com/user-attachments/assets/69d43884-afba-4418-b60d-ed1597ce352a

**Environment**
System:
  OS: macOS 14.6.1
  CPU: (10) arm64 Apple M1 Pro
  Memory: 418.08 MB / 32.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.17.0
    path: ~/.nvm/versions/node/v20.17.0/bin/node
  Yarn:
    version: 1.22.22
    path: ~/.nvm/versions/node/v20.17.0/bin/yarn
  npm:
    version: 10.8.2
    path: ~/.nvm/versions/node/v20.17.0/bin/npm
  Watchman:
    version: 2024.08.26.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /opt/homebrew/lib/ruby/gems/3.1.0/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 24.0
      - iOS 18.0
      - macOS 15.0
      - tvOS 18.0
      - visionOS 2.0
      - watchOS 11.0
  Android SDK: Not Found
IDEs:
  Android Studio: 2023.3 AI-233.14808.21.2331.11842104
  Xcode:
    version: 16.0/16A242d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.10
    path: /usr/bin/javac
  Ruby:
    version: 3.1.4
    path: /opt/homebrew/opt/ruby@3.1/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react: Not Found
  react-native: Not Found
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false

- expo version : 51
- using firebase : no
- using static build : yes
- routing : react-navigation
- package version (ex: 1.2.0) : 2.3.0
- platform target : iOS
- device : simulator
- type: local build

**Comment by achorein** [comment]
:/
any error in console or xcode logs ?

**Comment by mfkrause** [comment]
This is _likely_ due to this function: https://github.com/achorein/expo-share-intent/blob/1c41e641bc1e6c91999d10ba63477ab9f254430f/plugin/src/ios/ShareExtensionViewController.swift#L251-L269

In line 254, you're using an `openURL` selector on `UIResponder`. This has been deprecated for quite a while (past iOS 10) and has now been completely disabled in the iOS 18 SDK: https://developer.apple.com/documentation/uikit/uiapplication/1622961-openurl

However, it's not as simple as changing the selector in line 254 to i.e. `open:options:completionHandler:`. [Here's a write-up on a similar issue](https://keyboardkit.com/blog/2024/09/11/ios18-breaks-selector-based-url-opening) in the KeyboardKit SwiftUI SDK (unrelated package, but the issue is effectively the same).

Unfortunately, I currently don't have the time to propose a fix, but just wanted to leave this here in case it helps. For now, you'll need to build with the iOS 17 SDK / Xcode 15. If you do, the share extension will work fine in iOS 18, too. Currently, the Expo EAS cloud still uses Xcode 15, therefore it works fine if you build the app using EAS Cloud (once they switch to Xcode 16 / iOS 18 SDK, you might be able to force the old Xcode / SDK by using the [`macos-sonoma-14.5-xcode-15.4` server image](https://docs.expo.dev/build-reference/infrastructure/#macos-sonoma-145-xcode-154-latest-sdk-51-sdk-50-sdk-49)).

**Comment by cback97** [comment]
getting this same error when sharing an image, probably related.

**Comment by achorein** [comment]
I will check xcode16 and ios18 issues next week.

**Comment by darix795** [comment]
Ok, i tried to reproduce error in this repo and I solved by replacing `SLComposeServiceViewController` extension to `UIViewController` and `application.perform(selectorOpenURL, with: url) ` with `application.open(url)` in according with Apple's offical docs.

```diff
 - class ShareViewController: SLComposeServiceViewController {
 + class ShareViewController: UIViewController {
```

```diff
- application.perform(selectorOpenURL, with: url) 
+ application.open(url)
```

https://github.com/achorein/expo-share-intent/blob/1c41e641bc1e6c91999d10ba63477ab9f254430f/plugin/src/ios/ShareExtensionViewController.swift#L259

After this changes it seems to works fine on Xcode 16 and iOS 18. I was tested with links, files and images.
Unfortunately, i haven't permission to open new pull requests.

**Comment by achorein** [comment]
@darix795 thank you for digging this, i'm going to to test this on my own.
btw you can fork this repo and make a PR from it.

**Comment by achorein** [comment]
Fixed in v2.4.0
